flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala

```diff
@@ -79,7 +83,27 @@ abstract class Expand(
     litmus.succeed()
   }
 
-  override def deriveRowType(): RelDataType = outputRowType
+  override def deriveRowType(): RelDataType = {
+    val fieldNullableMap = mutable.Map[Int, Boolean]()
+    projects.map {
+      project =>
+        project.zipWithIndex.map {
+          f => {
+            val nullable = fieldNullableMap.get(f._2).getOrElse(f._1.getType.isNullable)
+            fieldNullableMap.put(f._2, nullable || f._1.getType.isNullable)
+          }
+        }
+    }
+    val typeList = ListBuffer[RelDataType]()
+    val typeFactory = cluster.getTypeFactory
+    projects.get(0).zipWithIndex.map {
```

Comment:
another case we should consider is Expand can be more generic (not only used for distinct agg, but also for other case), after that the fields in the Expand node may be different type, so we should check whether a column's types in all expanded rows are the same or we should derive the common type for each column.

Reply:
good point！I'll try to investigate the two ways.